// armstrong-react/dist/utilities/keyCodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var KeyCodes;
(function (KeyCodes) {
    KeyCodes[KeyCodes["backspace"] = 8] = "backspace";
    KeyCodes[KeyCodes["tab"] = 9] = "tab";
    KeyCodes[KeyCodes["enter"] = 13] = "enter";
    KeyCodes[KeyCodes["shift"] = 16] = "shift";
    KeyCodes[KeyCodes["ctrl"] = 17] = "ctrl";
    KeyCodes[KeyCodes["alt"] = 18] = "alt";
    KeyCodes[KeyCodes["pause_break"] = 19] = "pause_break";
    KeyCodes[KeyCodes["capsLock"] = 20] = "capsLock";
    KeyCodes[KeyCodes["escape"] = 27] = "escape";
    KeyCodes[KeyCodes["pageUp"] = 33] = "pageUp";
    KeyCodes[KeyCodes["pageDown"] = 34] = "pageDown";
    KeyCodes[KeyCodes["end"] = 35] = "end";
    KeyCodes[KeyCodes["home"] = 36] = "home";
    KeyCodes[KeyCodes["leftArrow"] = 37] = "leftArrow";
    KeyCodes[KeyCodes["upArrow"] = 38] = "upArrow";
    KeyCodes[KeyCodes["rightArrow"] = 39] = "rightArrow";
    KeyCodes[KeyCodes["downArrow"] = 40] = "downArrow";
    KeyCodes[KeyCodes["insert"] = 45] = "insert";
    KeyCodes[KeyCodes["delete"] = 46] = "delete";
    KeyCodes[KeyCodes["key_0"] = 48] = "key_0";
    KeyCodes[KeyCodes["key_1"] = 49] = "key_1";
    KeyCodes[KeyCodes["key_2"] = 50] = "key_2";
    KeyCodes[KeyCodes["key_3"] = 51] = "key_3";
    KeyCodes[KeyCodes["key_4"] = 52] = "key_4";
    KeyCodes[KeyCodes["key_5"] = 53] = "key_5";
    KeyCodes[KeyCodes["key_6"] = 54] = "key_6";
    KeyCodes[KeyCodes["key_7"] = 55] = "key_7";
    KeyCodes[KeyCodes["key_8"] = 56] = "key_8";
    KeyCodes[KeyCodes["key_9"] = 57] = "key_9";
    KeyCodes[KeyCodes["key_a"] = 65] = "key_a";
    KeyCodes[KeyCodes["key_b"] = 66] = "key_b";
    KeyCodes[KeyCodes["key_c"] = 67] = "key_c";
    KeyCodes[KeyCodes["key_d"] = 68] = "key_d";
    KeyCodes[KeyCodes["key_e"] = 69] = "key_e";
    KeyCodes[KeyCodes["key_f"] = 70] = "key_f";
    KeyCodes[KeyCodes["key_g"] = 71] = "key_g";
    KeyCodes[KeyCodes["key_h"] = 72] = "key_h";
    KeyCodes[KeyCodes["key_i"] = 73] = "key_i";
    KeyCodes[KeyCodes["key_j"] = 74] = "key_j";
    KeyCodes[KeyCodes["key_k"] = 75] = "key_k";
    KeyCodes[KeyCodes["key_l"] = 76] = "key_l";
    KeyCodes[KeyCodes["key_m"] = 77] = "key_m";
    KeyCodes[KeyCodes["key_n"] = 78] = "key_n";
    KeyCodes[KeyCodes["key_o"] = 79] = "key_o";
    KeyCodes[KeyCodes["key_p"] = 80] = "key_p";
    KeyCodes[KeyCodes["key_q"] = 81] = "key_q";
    KeyCodes[KeyCodes["key_r"] = 82] = "key_r";
    KeyCodes[KeyCodes["key_s"] = 83] = "key_s";
    KeyCodes[KeyCodes["key_t"] = 84] = "key_t";
    KeyCodes[KeyCodes["key_u"] = 85] = "key_u";
    KeyCodes[KeyCodes["key_v"] = 86] = "key_v";
    KeyCodes[KeyCodes["key_w"] = 87] = "key_w";
    KeyCodes[KeyCodes["key_x"] = 88] = "key_x";
    KeyCodes[KeyCodes["key_y"] = 89] = "key_y";
    KeyCodes[KeyCodes["key_z"] = 90] = "key_z";
    KeyCodes[KeyCodes["leftWindowKey"] = 91] = "leftWindowKey";
    KeyCodes[KeyCodes["rightWindowKey"] = 92] = "rightWindowKey";
    KeyCodes[KeyCodes["selectKey"] = 93] = "selectKey";
    KeyCodes[KeyCodes["numpad0"] = 96] = "numpad0";
    KeyCodes[KeyCodes["numpad1"] = 97] = "numpad1";
    KeyCodes[KeyCodes["numpad2"] = 98] = "numpad2";
    KeyCodes[KeyCodes["numpad3"] = 99] = "numpad3";
    KeyCodes[KeyCodes["numpad4"] = 100] = "numpad4";
    KeyCodes[KeyCodes["numpad5"] = 101] = "numpad5";
    KeyCodes[KeyCodes["numpad6"] = 102] = "numpad6";
    KeyCodes[KeyCodes["numpad7"] = 103] = "numpad7";
    KeyCodes[KeyCodes["numpad8"] = 104] = "numpad8";
    KeyCodes[KeyCodes["numpad9"] = 105] = "numpad9";
    KeyCodes[KeyCodes["multiply"] = 106] = "multiply";
    KeyCodes[KeyCodes["add"] = 107] = "add";
    KeyCodes[KeyCodes["subtract"] = 109] = "subtract";
    KeyCodes[KeyCodes["decimalPoint"] = 110] = "decimalPoint";
    KeyCodes[KeyCodes["divide"] = 111] = "divide";
    KeyCodes[KeyCodes["f1"] = 112] = "f1";
    KeyCodes[KeyCodes["f2"] = 113] = "f2";
    KeyCodes[KeyCodes["f3"] = 114] = "f3";
    KeyCodes[KeyCodes["f4"] = 115] = "f4";
    KeyCodes[KeyCodes["f5"] = 116] = "f5";
    KeyCodes[KeyCodes["f6"] = 117] = "f6";
    KeyCodes[KeyCodes["f7"] = 118] = "f7";
    KeyCodes[KeyCodes["f8"] = 119] = "f8";
    KeyCodes[KeyCodes["f9"] = 120] = "f9";
    KeyCodes[KeyCodes["f10"] = 121] = "f10";
    KeyCodes[KeyCodes["f11"] = 122] = "f11";
    KeyCodes[KeyCodes["f12"] = 123] = "f12";
    KeyCodes[KeyCodes["numLock"] = 144] = "numLock";
    KeyCodes[KeyCodes["scrollLock"] = 145] = "scrollLock";
    KeyCodes[KeyCodes["semicolon"] = 186] = "semicolon";
    KeyCodes[KeyCodes["equalSign"] = 187] = "equalSign";
    KeyCodes[KeyCodes["comma"] = 188] = "comma";
    KeyCodes[KeyCodes["dash"] = 189] = "dash";
    KeyCodes[KeyCodes["period"] = 190] = "period";
    KeyCodes[KeyCodes["forwardSlash"] = 191] = "forwardSlash";
    KeyCodes[KeyCodes["graveAccent"] = 192] = "graveAccent";
    KeyCodes[KeyCodes["openBracket"] = 219] = "openBracket";
    KeyCodes[KeyCodes["backslash"] = 220] = "backslash";
    KeyCodes[KeyCodes["closeBracket"] = 221] = "closeBracket";
    KeyCodes[KeyCodes["singleQuote"] = 222] = "singleQuote";
})(KeyCodes = exports.KeyCodes || (exports.KeyCodes = {}));
